capsules/src/buzzer.rs

```diff
@@ -0,0 +1,143 @@
+//! Provides userspace access to a buzzer.
+//!
+//! ## Instantiation
+//!
+//! Instantiate the capsule for use as a syscall driver, using the corresponding service capsule.
+//! For example, using the pwm buzzer:
+//!
+//! ``` rust
+//!
+//! let buzzer = static_init!(
+//!     capsules::buzzer::Buzzer<'static>,
+//!     capsules::buzzer::Buzzer::new(
+//!         pwm_buzzer,
+//!         board_kernel.create_grant(capsules::buzzer::DRIVER_NUM, &memory_allocation_capability)
+//!     )
+//! );
+//!
+//! ```
+
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::OptionalCell;
+use kernel::{ErrorCode, ProcessId};
+
+// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Buzzer as usize;
+
+pub struct Buzzer<'a, B: hil::buzzer::Buzzer<'a>> {
+    /// The service capsule buzzer.
+    buzzer: &'a B,
+    /// Per-app state.
+    apps: Grant<(), UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    /// Which app is currently using the buzzer.
+    active_app: OptionalCell<ProcessId>,
+}
+
+impl<'a, B: hil::buzzer::Buzzer<'a>> Buzzer<'a, B> {
+    pub fn new(
+        buzzer: &'a B,
+        grant: Grant<(), UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> Buzzer<'a, B> {
+        Buzzer {
+            buzzer: buzzer,
+            apps: grant,
+            active_app: OptionalCell::empty(),
+        }
+    }
+
+    pub fn is_valid_app(&self, appid: ProcessId) -> bool {
```

Comment:
Please add a comment to explain what a valid app means for this driver.

Reply:
Done